core/src/storage2/alloc/init.rs

```diff
@@ -0,0 +1,160 @@
+    Deploy,
+    /// Initializes the global dynamic storage allocator from storage.
+    ///
+    /// Upon initialization the dynamic storage allocator will be pulled
+    /// from the contract storage with the assumption that a former
+    /// contract deployment has already taken place in the past.
+    Call,
+}
+
+/// The state of the dynamic allocator global instance.
+#[derive(Debug)]
+#[allow(clippy::large_enum_variant)]
+enum DynamicAllocatorState {
+    /// The global instance has not yet been initialized.
+    ///
+    /// Upon initialization it will be created from scratch as if the
+    /// contract has been deployed for the first time.
+    UninitDeploy,
+    /// The global instance has not yet been initialized.
+    ///
+    /// Upon initialization it will be pulled from the contract storage
+    /// with the assumption that a former contract deployment has already
+    /// taken place in the past.
+    UninitCall,
+    /// The global instance has already been initialized successfully.
+    Initialized(DynamicAllocator),
+}
+
+impl From<ContractPhase> for DynamicAllocatorState {
+    fn from(phase: ContractPhase) -> Self {
+        match phase {
+            ContractPhase::Deploy => DynamicAllocatorState::UninitDeploy,
+            ContractPhase::Call => DynamicAllocatorState::UninitCall,
+        }
+    }
+}
+
+cfg_if! {
+    if #[cfg(all(not(feature = "std"), target_arch = "wasm32"))] {
+        // Procedures for the Wasm compilation:
+
+        /// The global instance for the dynamic storage allocator.
+        static mut GLOBAL_INSTANCE: DynamicAllocatorState = DynamicAllocatorState::UninitDeploy;
+
+        /// Commands the (re-)initialization of the global instance for the dynamic
+        /// storage allocator.
+        pub fn initialize_for(phase: ContractPhase) {
+            let instance = unsafe { &mut GLOBAL_INSTANCE };
+            // We do not allow reinitialization for Wasm targets for performance reasons.
+            if let DynamicAllocatorState::Initialized(_) = instance {
+                panic!("cannot reinitialize dynamic storage allocator instance in Wasm");
+            }
+            *instance = phase.into();
+        }
+
+        /// Runs the given closure on the global instance for the dynamic storage allocator.
+        pub fn on_call<F, R>(f: F) -> R
+        where
+            F: FnOnce(&mut DynamicAllocator) -> R,
+        {
+            let instance = unsafe { &mut GLOBAL_INSTANCE };
+            match instance {
+                DynamicAllocatorState::UninitDeploy => {
+                    let mut allocator = DynamicAllocator::default();
+                    let result = f(&mut allocator);
+                    *instance = DynamicAllocatorState::Initialized(allocator);
+                    result
+                }
+                DynamicAllocatorState::UninitCall => {
+                    let mut allocator = pull_spread_root::<DynamicAllocator>(&Key(DYNAMIC_ALLOCATOR_KEY_OFFSET));
```

Comment:
Where's the equivalent `push_spread_root` for the DynamicAllocator?

Reply:
in its implied `Drop` implementation and also this has to be guided by `ink_lang` later.

Reply:
I can't find the `Drop` implementation which does this - just trying to understand the lifecycle of the allocator

Reply:
The allocator doesn't require an explicit `Drop` implementation and so doesn't its internal `storage::BitStash` but still their `Drop` impls forward implicitely to the `Drop` impls of their fields. So what gets dropped in the end is a `storage::Bitvec` and a `storage::Vec`.

Reply:
The life cycle of the dynamic storage allocator is different in `std` and `Wasm` compilation for reasons of making it better testable in `std` (off-chain) environments. In Wasm compilation we know there is only one thread and we construct a dynamic storage allocator within a `ManuallyDrop` at the start but before that we tell it how to construct:
- Either for `construct` or for `call`. Upon `contsruct` we create a brand new dynamic storage allocator because we cannot rely on having one on the storage already. And upon `call` we simply pull the one that has to be there from the `construct`.

Reply:
But `Drop` in these cases will clear, not push - where does the `push` happen?

Reply:
The `push` will be managed by the `ink_lang` side.

Reply:
We will add a `PushOnDrop` abstraction in a follow-up PR that fixes this problem. This abstraction will also be useful for the static storage of a smart contract.